# Labor 3

A párhuzamos algoritmusok célja a [teljesítmény növelése](https://courses.cs.washington.edu/courses/csep524/07sp/poppChaper1.pdf). Ezt azzal érik el, hogy a számításokat egyszerre több számítógépen vagy számítási egységen végzik el, így csökkentve a számítási időt, vagy lehetővé téve pontosabb eredmények elérését egy adott időegység alatt.


# Teljesítménymértékek

A teljesítménynövekedés számszerű kifejezése érdekében áttekintjük, hogy milyen módszerek állnak rendelkezésre.

## Gyorsítás (speedup)

Az időarány, amelyben a párhuzamos algoritmus végrehajtása gyorsabb, mint a szekvenciális algoritmusé. A gyorsítás értéke az eredeti, szekvenciális idő és az új idő hányadosa.

Jelölje $T(n,p)$ azt a függvényt, ami megadja, hogy egy párhuzamos program egy $n$ méretű feladatot mennyi idő alatt old meg $p$ processzor használatával.

Adott $n$-méretű feladatra a $T(n,p)$ párhuzamos $p$ processzoron futó program reális  **gyorsítását (gyorsítási faktort)**

$S(n,p)=\frac{T^*(n)}{T(n,p)}$

adja meg, ahol $T^*(n)$ a leggyorsabb ismert soros algoritmus futási ideje.



A környezettől függően a gyorsítás vonatkozhat $p$ folyamat vagy $p$ szál alkalmazására is.

$T^*(n)$ nem mindig ismert, vagy nem implementálható, ezért gyakran a relatív **_gyorsítását_**  használjuk, ahol a párhuzamos program egy processzoros futási idejét vesszük figyelembe:

$S(n,p)=\frac{T(n,1)}{T(n,p)}$


## Hatékonyság (efficiency)

Másik fontos metrika a **_párhuzamos hatékonyság_** ami kifejezi a gyorsítás arányát a felhasznált számítási erőforrásokhoz képest. Az hatékonyság értéke a gyorsulás és a felhasznált erőforrások (például a processzorok vagy a magok száma) hányadosa:

$E(n,p)=\frac{S(n,p)}{p}=\frac{T^*(n)}{p*T(n,p)}$

Ez a mutató betekintést nyújt a processzorok kihasználtságára, ideális esetben ez az érték 1 (lineáris gyorsulás).

Egy párhuzamos programot akkor nevezünk **_skálázhatónak_**, ha a párhuzamos hatékonyság fenntartható. Skálázhatóság (scalability): a rendszer képessége, hogy a számítási feladatokat nagyobb erőforrásokkal bővítve hatékonyan kezelje.



# Időmérésa CUDA események segítségével

A CUDA esemény ([CUDA event](https://docs.nvidia.com/cuda/cuda-runtime-api/index.html#group__CUDART__EVENT)) egy szinkronizációs jel, amely lehetővé teszi a programozó számára, hogy monitorizálja a CUDA munkafolyamatokat, szinkronizálja a CUDA adatfolyamokat (CUDA streams) és lehetőve teszi a [CUDA kernelek pontos időmérését](https://developer.nvidia.com/blog/how-implement-performance-metrics-cuda-cc/).


A `cudaEvent` egy CUDA API függvény, amely lehetővé teszi a programozók számára, hogy mérjék a CUDA műveletek végrehajtási idejét. A `cudaEvent` funkciók segítségével időbélyegeket állíthatunk be, majd ezen bélyegek között eltelt időt mérhetjük.

A következő lépésekkel használjuk a `cudaEvent`-et időmérésre:

1. Létre kell hozni két `cudaEvent` objektumot, amelyek közül az egyiket a művelet kezdeteként, a másikat pedig a művelet befejezéseként jelöljük meg. Például:
  ```cpp
  cudaEvent_t start, stop;
  cudaEventCreate(&start);
  cudaEventCreate(&stop);
  ```
2. A [`cudaEventRecord`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1gf4fcb74343aa689f4159791967868446) függvény segítségével be kell állítanunk a start és stop bélyegeket a művelet elkezdése elött és a befejezésnél. Például:
  ```cpp
  cudaEventRecord(start, 0);
  // CUDA műveletek végrehajtása
  // pl. cudaMalloc, cudaMemcpy, kernel hívás, eredmények visszamásolása, memória felszabadítás
  cudaEventRecord(stop, 0);
  ```
  A második paraméter a `cudaEventRecord` hívásokban az adatfolyamot ([CUDA stream](https://developer.download.nvidia.com/CUDA/training/StreamsAndConcurrencyWebinar.pdf)) jelöli, és a 0 érték azt jelzi, hogy az esemény a null adatfolyamhoz van rendelve.

  Az adatfolyamok (stream) olyan CUDA fogalmak, amelyek lehetővé teszik a kernel futtatását és az adatmozgatást aszinkron módon. Az adatfolyamok segítségével párhuzamosíthatjuk a műveleteket és jobban kihasználhatjuk a hardverünk párhuzamosítási képességeit.

3. A [`cudaEventSynchronize`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g949aa42b30ae9e622f6ba0787129ff22) függvény használatával megvárjuk, amíg a stop időbélyeg rögzítése megtörténik, majd ellenőrizzük az esetleges hibákat. Például:
  ```cpp
  cudaEventSynchronize(stop);
  cudaError_t error = cudaGetLastError();
  if (error != cudaSuccess) {
      printf("CUDA hiba: %s\n", cudaGetErrorString(error));
  }
  ```

4. Végül az [`cudaEventElapsedTime`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g40159125411db92c835edb46a0989cd6) függvény segítségével kiszámítjuk a két bélyeg között eltelt időt. Például:
  ```cpp
  float elapsedTime;
  cudaEventElapsedTime(&elapsedTime, start, stop);
  printf("A CUDA műveletek elvégzése %.2f ms telt.\n", elapsedTime);
  ```
5. A [`cudaEventDestroy`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html#group__CUDART__EVENT_1g2cb6baa0830a1cd0bd957bfd8705045b) függvényt használjuk az általunk létrehozott `cudaEvent` objektumok felszabadítására. Például:
  ```cpp
  cudaEventDestroy(start);
  cudaEventDestroy(stop);
  ```

Az alábbi példa összefoglalja a fenti lépéseket:

```cpp
__global__ void addKernel(int *c, const int *a, const int *b, int size)
{
    int i = threadIdx.x;
    if (i < size) {
        c[i] = a[i] + b[i];
    }
}

int main()
{
    const int size = 1024;
    int a[size], b[size], c[size];

    for (int i = 0; i < size; i++) {
        a[i] = i;
        b[i] = 2 * i;
    }

    int *dev_a, *dev_b, *dev_c;

    cudaEvent_t start, stop;
    cudaEventCreate(&start);
    cudaEventCreate(&stop);

    cudaEventRecord(start, 0);

    cudaMalloc(&dev_a, size * sizeof(int));
    cudaMalloc(&dev_b, size * sizeof(int));
    cudaMalloc(&dev_c, size * sizeof(int));

    cudaMemcpy(dev_a, a, size * sizeof(int), cudaMemcpyHostToDevice);
    cudaMemcpy(dev_b, b, size * sizeof(int), cudaMemcpyHostToDevice);

    addKernel<<<1, size>>>(dev_c, dev_a, dev_b, size);

    cudaMemcpy(c, dev_c, size * sizeof(int), cudaMemcpyDeviceToHost);
    
    cudaFree(dev_a);
    cudaFree(dev_b);
    cudaFree(dev_c);

    cudaEventRecord(stop, 0);
    cudaEventSynchronize(stop);

    cudaError_t error = cudaGetLastError();
    if (error != cudaSuccess) {
        printf("CUDA hiba: %s\n", cudaGetErrorString(error));
    }
    float elapsedTime;
    cudaEventElapsedTime(&elapsedTime, start, stop);
    printf("A CUDA művelet %.2f millió másodpercig tartott.\n", elapsedTime);
    
    cudaEventDestroy(start);
    cudaEventDestroy(stop);
    return 0;
}
```

## Időmérés CPU-n

C++-ban, egy nagyfelbontású aktuális időpontot, a
```cpp
static  std::chrono::time_point<std::chrono::high_resolution_clock> now()  noexcept;
```
[függvénnyel](https://en.cppreference.com/w/cpp/chrono/high_resolution_clock/now) kérhetjük le. A rendszer maximális időfelbontását (_ticks per second_) a `std::chrono::high_resolution_clock::period::den` érték lekérésével kaphatjuk meg.

Ha kíváncsiak vagyunk egy kódrészlet futási idejére, megmérhetjük azt két időbélyegek segítségével, melyeket lekérdezünk mindjárt a kódrészlet előtt és után. A két időbélyeg különbsége megadja az eltelt időt, a kívánt [mértékegységben](https://en.cppreference.com/w/cpp/chrono/duration/duration_cast):

![](https://i.ibb.co/F5D8brL/duration-cast.png)


### Példa időmérésre:

```cpp
#include <iostream>
#include <chrono>

using  namespace std;

int main()
{
	cout  << chrono::high_resolution_clock::period::den  << endl;
	auto start_time = chrono::high_resolution_clock::now();
	long temp =  0;
	for  (auto i =  0; i < 100000000; i++)
		temp += i;
	auto end_time = chrono::high_resolution_clock::now();
	cout  << chrono::duration_cast<chrono::seconds>(end_time - start_time).count()  <<  ":";
	cout  << chrono::duration_cast<chrono::milliseconds>(end_time - start_time).count()  <<  ":";
	cout  << chrono::duration_cast<chrono::microseconds>(end_time - start_time).count()  <<  ":";
	cout << endl;

	return  0;
}
```
Mivel ez az időmérés a rendszerórát veszi figyelembe (wall clock time), és a számítások közben más folyamatok is ütemezésre kerülhettek, a pontos eredmény érdekében érdemes többször megmérni a futási időt, az kiugró értékeket eldobni, majd a többi mérésből átlagot számolni.

A `clock_gettime` segítségével, POSIX kompatibilis rendszereken a `CLOCK_PROCESS_CPUTIME_ID` óra azonosító paraméter megadásával, folyamat szintű időmérést végezhetünk. A [https://en.cppreference.com/w/c/chrono/clock](https://en.cppreference.com/w/c/chrono/clock) található példa bemutatja a folyamat és rendszer idő közötti különbséget.

# Véletlenszám generálás CUDA-ban

A továbbiakban megismerkedünk a véletlen számok generálásának módszertanával CUDA-ban, a [`cuRAND`](https://docs.nvidia.com/cuda/curand/index.html)  könyvtár használatával.


Egy álvéletlenszám sorozat generálása egy szekvenciális folyamat, ahol a következő számot az előbbi segítségével állítjuk elő. Például a [lineáris kongruenciális generátor](https://wikihuhu.top/wiki/Lehmer_random_number_generator) esetében


```
X_{n+1} = (aX_{n} + c) mod m
```

ahol ![](https://render.githubusercontent.com/render/math?math=$a$) a szorzó , ![](https://render.githubusercontent.com/render/math?math=$c$) a növekmény és ![](https://render.githubusercontent.com/render/math?math=$m$) a modulus.

A kezdeti kifejezést magnak (angolul seed) nevezzük. Ez teszi lehetővé egy látszólag véletlenszerű szekvencia létrehozását. Minden maghoz új folytatást kapunk.

Párhuzamosan generálni pszeudo-véletlen sorozattokat [kicsit bonyolultabb](https://quick-adviser.com/is-rand--thread-safe/
), mivel egy naiv megközelítésben az `X_{n+1}` kiszámolásánál versenyhelyzet lép fel, több szál is ugyanazt  az értéket fogja generálni.


Egy egyszerű megoldás, ha  minden szálhoz külön véletlenszerű állapotot hozunk létre, amelyet az egyes szálak a saját, a többiektől független véletlenszerű számsorozat létrehozásához használnak. Minden szál különböző magból, eredeti állapotból indul ki, annak érdekében, hogy mindegyikük egy különböző álváletlenszerű sorozatot generáljon.

A [`cuRAND`](https://docs.nvidia.com/cuda/curand/index.html) könyvtár interfészt biztosít a véletlenszám-generátor állapotának inicializálásához szálanként, és ennek az állapotnak a felhasználásához véletlenszám-sorozatok generálásához. A következő lépések szükségesek ahhoz, hogy a cuRAND segítségével véletlenszerű számokat tudjunk generálni a CUDA programjainkban:

1.	A `cuRAND` fejlécek behívása.

	```cpp
	#include <curand_kernel.h>
	#include <curand.h>
	```
2.	Minden szálnak külön kell biztosítani egy eredeti magot,  különböző belső állapotot:

	-	Memória lefoglalása a szálak állapváltozóinak (`curandState`).

		```cpp
		curandState *dev_random;
		cudaMalloc((void**)&dev_random, num_threads_per_block*num_blocks*sizeof(curandState));
		```

	- Véletlenszerű állapotok inicializálása egy kernelben. Minden szál inicializálja a saját állapotát a GPU-n, a saját egyedi azonosítóját felhasználva:

		```cpp
		__global__ void gpu_random(..., curandState *states) {  
			int id = threadIdx.x + blockDim.x * blockIdx.x;  

			...  
			int seed = id; // different seed per thread         
			curand_init(seed, id, 0, &states[id]);  //  Initialize CURAND        
		 	...
		 ```

3. Véletlen számok generálása a GPU-n valamilyen eloszlás szerint (pl. `curand_uniform`), szálanként a saját inicializált állapot használatával.

	```cpp
	__global__ void gpu_random(..., curandState *states) {  
		int id = threadIdx.x + blockDim.x * blockIdx.x;  float x;         
		...  
		curand_init(seed, tid, 0, &states[id]);  //  Initialize CURAND         
		...
		for(int i = 0; i < TRIALS_PER_THREAD; i++) {   
		x = curand_uniform (&states[id]);   
		...  
	}
	```


A leggyakrabban használt eloszlások:

```cpp
// Egyenletes eloszlás
__device__ float
curand_uniform (curandState_t *state)

// Normális eloszlás
__device__ float
curand_normal (curandState_t *state)

// Log-normális eloszlás - a valószínűségi változó logaritmusa normális eloszlású.
__device__ float
curand_log_normal (curandState_t *state, float mean, float stddev)

// Poisson-eloszlás
__device__ unsigned int
curand_poisson (curandState_t *state, double lambda)
```

Ezek a függvények elérhetőek dupla pontosságú lebegőpontos (`double`) számábrázolással is.

## [Példa](https://github.com/deeperlearning/professional-cuda-c-programming/blob/master/examples/chapter08/rand-kernel.cu)

```cpp
#include <stdio.h>
#include <stdlib.h>
#include <cuda.h>
#include <curand_kernel.h>


int threads_per_block = 256;
int blocks_per_grid = 30;


/*
 * device_api_kernel uses the cuRAND device API to generate random numbers
 * on-the-fly on the GPU, and then performs some dummy computation using them.
 */
__global__ void device_api_kernel(curandState *states, float *out, int N)
{
    int i;
    int tid = blockIdx.x * blockDim.x + threadIdx.x;
    int nthreads = gridDim.x * blockDim.x;
    curandState *state = states + tid;

    curand_init(9384, tid, 0, state);

    for (i = tid; i < N; i += nthreads)
    {
        float rand = curand_uniform(state);
        rand = rand * 2;
        out[i] = rand;
    }
}


/*
 * use_device_api is an examples usage of the cuRAND device API to use the GPU
 * to generate random values on the fly from inside a CUDA kernel.
 */
void use_device_api(int N)
{
    int i;
    static curandState *states = NULL;
    float *dOut, *hOut;

    /*
     * Allocate device memory to store the output and cuRAND device state
     * objects (which are analogous to handles, but on the GPU).
     */
    cudaMalloc((void **)&dOut, sizeof(float) * N);
    cudaMalloc((void **)&states, sizeof(curandState) *
                threads_per_block * blocks_per_grid);
    hOut = (float *)malloc(sizeof(float) * N);

    // Execute a kernel that generates and consumes its own random numbers
    device_api_kernel<<<blocks_per_grid, threads_per_block>>>(states, dOut, N);

    // Retrieve the results
    cudaMemcpy(hOut, dOut, sizeof(float) * N, cudaMemcpyDeviceToHost);

    printf("Sampling of output from device API:\n");

    for (i = 0; i < 10; i++)
    {
        printf("%2.4f\n", hOut[i]);
    }

    printf("...\n");

    free(hOut);
    cudaFree(dOut);
    cudaFree(states);
}

int main(int argc, char **argv)
{
    int N = 8388608;

    use_device_api(N);

    return 0;
}
```



#CUDA profilozása az NVIDIA Profiler segítségével

A CUDA-programok hatékonyságának elemzéséhez gyakran nem elég csupán a kernel futási időt mérni, mivel ez nem ad részletes információt a **memóriahozzáférésekről, regiszterhasználatról, párhuzamos végrehajtás hatékonyságáról és az adatmozgatásokról**. Az [NVIDIA Profiler eszközei](https://developer.nvidia.com/performance-analysis-tools), például **nvprof, Nsight Systems és Nsight Compute**, lehetőséget adnak mélyebb teljesítményprofilozásra.  

**Mivel más vagy több  az NVIDIA Profiler, mint egy egyszerű időmérés?**  
- Megmutatja, hogy a kernel futásán belül hol vannak szűk keresztmetszetek.  
- Elemzi a **memóriahasználatot** (globális, megosztott memória, L2 cache kihasználtság).  
- Vizsgálja a **számítási kihasználtságot** (SM-kihasználtság, párhuzamos végrehajtás).  
- Részletes jelentést nyújt az egyes CUDA API hívásokról és memóriamásolásokról.  

Az [nvprof](https://docs.csc.fi/computing/nvprof/) eszköz egy parancssori profiler, nem igényel GUI-t, ezért Google Colab környezetben is könnyen tudjuk használni.


Az nvprof-nak csak egyszerűen át kell adnunk a CUDA programunkat mint paraméter:

```bash
nvprof ./cuda_program
```

Ez majd automatikusan rögzíti az összes CUDA API hívást, kernel futtatást és memóriaátvitelt, majd egy összegző jelentést készít.

---

## Példa -  CUDA program profilozása
Vizsgáljuk meg egy egyszerű CUDA-kernel futását és memóriahasználatát az nvprof segítségével.


In [ ]:
%%writefile profiletest.cu

#include <cuda.h>
#include <curand_kernel.h>
#include <stdio.h>

__global__ void vector_add(float *a, float *b, float *c, int n, unsigned long long seed) {
    int i = threadIdx.x + blockIdx.x * blockDim.x;

    // Véletlenszám-generátor inicializálása minden szálnál
    curandState state;
    curand_init(seed, i, 0, &state);

    if (i < n) {
        float noise = curand_uniform(&state) * 0.1f - 0.05f; // Zaj [-0.05, 0.05] intervallumban
        c[i] = a[i] + b[i] + noise;
    }
}

int main() {
    int n = 1000000;
    size_t size = n * sizeof(float);

    // Memória foglalás a hoston
    float *h_a = (float*)malloc(size);
    float *h_b = (float*)malloc(size);
    float *h_c = (float*)malloc(size);

    // Inicializálás
    for (int i = 0; i < n; i++) {
        h_a[i] = 1.0f;
        h_b[i] = 2.0f;
    }

    // Memória foglalás a GPU-n
    float *d_a, *d_b, *d_c;
    cudaMalloc(&d_a, size);
    cudaMalloc(&d_b, size);
    cudaMalloc(&d_c, size);

    // Adatok másolása a GPU-ra
    cudaMemcpy(d_a, h_a, size, cudaMemcpyHostToDevice);
    cudaMemcpy(d_b, h_b, size, cudaMemcpyHostToDevice);

    // Kernel indítása
    int blockSize = 256;
    int numBlocks = (n + blockSize - 1) / blockSize;
    vector_add<<<numBlocks, blockSize>>>(d_a, d_b, d_c, n, 1234ULL);

    // Eredmények visszamásolása
    cudaMemcpy(h_c, d_c, size, cudaMemcpyDeviceToHost);

    // Memória felszabadítás
    cudaFree(d_a); cudaFree(d_b); cudaFree(d_c);
    free(h_a); free(h_b); free(h_c);

    return 0;
}



Overwriting profiletest.cu


In [ ]:
!nvcc -arch=sm_75 -o vector_add_noise profiletest.cu

In [ ]:
!nvprof --version

nvprof: NVIDIA (R) Cuda command line profiler
Copyright (c) 2012 - 2024 NVIDIA Corporation
Release version 12.5.82 (21)


In [ ]:
!nvidia-smi

Tue Mar 25 13:38:44 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 550.54.15              Driver Version: 550.54.15      CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   50C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [ ]:
!nvprof ./vector_add_noise

==2094== NVPROF is profiling process 2094, command: ./vector_add_noise
==2094== Profiling application: ./vector_add_noise
==2094== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   92.88%  40.913ms         1  40.913ms  40.913ms  40.913ms  vector_add(float*, float*, float*, int, __int64)
                    3.91%  1.7237ms         1  1.7237ms  1.7237ms  1.7237ms  [CUDA memcpy DtoH]
                    3.21%  1.4135ms         2  706.77us  694.38us  719.15us  [CUDA memcpy HtoD]
      API calls:   69.89%  108.99ms         3  36.330ms  77.995us  108.83ms  cudaMalloc
                   29.34%  45.762ms         3  15.254ms  897.51us  43.957ms  cudaMemcpy
                    0.36%  566.31us         3  188.77us  151.47us  208.06us  cudaFree
                    0.30%  469.98us         1  469.98us  469.98us  469.98us  cudaLaunchKernel
                    0.09%  143.06us       114  1.2540us     103ns  57.994us  cuDeviceGetAttribu

**Mit tudunk meg ebből?**
- A **vector_add kernel** a teljes futási idő **92.15%-át** teszi ki.
- Az **adatmásolás** (Host → Device és Device → Host) mennyi időt vesz igénybe (~4.47% + 3.38%).
- Milyen API hívások hajtodtak végre.


## Részletesebb Elemzés
Az nvprof további hasznos lehetőségeket biztosít. A GPU-műveletek részletes megjelenítése hasnzáljuk:


In [ ]:
!nvprof --print-gpu-trace ./vector_add_noise

==2823== NVPROF is profiling process 2823, command: ./vector_add_noise
==2823== Profiling application: ./vector_add_noise
==2823== Profiling result:
   Start  Duration            Grid Size      Block Size     Regs*    SSMem*    DSMem*      Size  Throughput  SrcMemType  DstMemType           Device   Context    Stream  Name
243.29ms  696.81us                    -               -         -         -         -  3.8147MB  5.3462GB/s    Pageable      Device     Tesla T4 (0)         1         7  [CUDA memcpy HtoD]
244.22ms  707.98us                    -               -         -         -         -  3.8147MB  5.2619GB/s    Pageable      Device     Tesla T4 (0)         1         7  [CUDA memcpy HtoD]
245.41ms  40.934ms           (3907 1 1)       (256 1 1)        61        0B        0B         -           -           -           -     Tesla T4 (0)         1         7  vector_add(float*, float*, float*, int, __int64) [130]
286.35ms  1.6342ms                    -               -         -        

```bash
nvprof --print-gpu-trace ./vector_add
```

A profilozás az alkalmazás futtatásának különböző aspektusait méri, mint például a memória másolásokat, kernel futtatásokat, és a GPU erőforrások használatát. A profilozás minden CUDA művelet pontos időbélyegét megmutatja. Pontosan mit láthatunk?


### Értelmezés

1. CUDA Memcpy HtoD (Host-to-Device Memory Copy)
- **`243.29ms`** és **`244.22ms`**: Ez az időpont, amikor a memória másolás elkezdődik és befejeződik. A másolás a host (CPU) és a device (GPU) között történik.
- **`696.81us` és `707.98us`**: A memória másolás (Host to Device) ideje, tehát mennyi időbe telt az adatokat átmásolni a CPU-ról a GPU-ra.
- **`3.8147MB`**: A memória másolás mérete. Ez azt jelenti, hogy az adatok mérete körülbelül 3.8 MB volt, amit átvittünk.
- **`5.3462GB/s` és `5.2619GB/s`**: A memória másolás sebessége, azaz hány GB adatot másoltunk másodpercenként.

A **`SrcMemType`** és **`DstMemType`** az adatok forrás- és célmemóriájának típusát jelzi.
- **`Pageable`**: Ez azt jelenti, hogy a memóriát nem tartja fixen a rendszer, így a memória elérhetősége rugalmas.
- **`Device`**: A célmemória a GPU memóriája.


2. **CUDA Memcpy DtoH (Device-to-Host Memory Copy)**
- **`286.35ms`** és **`1.6342ms`**: A másolás befejeződése és az időtartam. Ez az adat visszamásolását jelenti a GPU-ról a CPU-ra.
- **`3.8147MB`**: A visszamásolt adat mennyisége.
- **`2.2796GB/s`**: Az adatok átvitelének sebessége a GPU-ról a CPU-ra.


3. **Kernel futás: `vector_add`**
- **`245.41ms`**: A kernel végrehajtásának időtartama, ami a `vector_add` nevű függvény futtatása.
- **`40.934ms`**: Ez az idő, amely alatt a kernel ténylegesen végrehajtódott.
- **`(3907 1 1)`**: A **grid mérete** (háromdimenziós), ami azt jelzi, hogy hány blokkot futtatott a kernel: 3907 blokkot.
- **`(256 1 1)`**: A **block mérete** (háromdimenziós), ami azt jelzi, hogy hány szálat futtatott egy blokkban: 256 szál.
- **`61`**: A kernel által használt **regiszterek** száma szálanként. A regiszterek a CPU/GPU gyors memóriája, amely a szálak gyors adatkezeléséhez szükségesek.
- **`0B`**: **Static shared memory (SSMem)** és **Dynamic shared memory (DSMem)**: Ez a memória, amelyet a blokk szálai osztanak meg egymással. Most nem lett használva sem statikus, sem dinamikus memória.
  
- A kernel neve: **`vector_add(float*, float*, float*, int, __int64)`**, ami azt jelzi, hogy egy vektorműveletet hajtott végre, ahol három lebegőpontos számokból álló vektort adtak össze.



### Következtetés

- Az adatátvitel (Host-to-Device és Device-to-Host) viszonylag gyors volt, 5-5 GB/s sebességgel.
- A **`vector_add`** kernel futása 245.41ms-ig tartott, és az átlagos végrehajtási idő 40.934ms volt. Az adatokat 3907 blokkban és 256 szálat használva dolgozták fel.
- A regiszterek száma szálanként 61, ami azt jelzi, hogy nem használtak túl sok regisztert a kernelben. A kernel nem használt megosztott memóriát (sem statikus, sem dinamikus).

Mint látjuk, a profilozás segít abban, hogy áttekinthessük a program GPU erőforrás-használatát, és láthatjuk, hogy a memória másolás és a kernel végrehajtás hogyan befolyásolja a program teljesítményét.




### Memóriahasználat Profilozása

In [ ]:
!nvprof --print-gpu-summary ./vector_add_noise

==4856== NVPROF is profiling process 4856, command: ./vector_add_noise
==4856== Profiling application: ./vector_add_noise
==4856== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   92.87%  40.918ms         1  40.918ms  40.918ms  40.918ms  vector_add(float*, float*, float*, int, __int64)
                    3.68%  1.6196ms         1  1.6196ms  1.6196ms  1.6196ms  [CUDA memcpy DtoH]
                    3.45%  1.5220ms         2  761.00us  743.66us  778.35us  [CUDA memcpy HtoD]


Az **`nvprof --print-gpu-summary`** parancs összegző profilozási adatokat adott vissza a CUDA programról, beleértve a GPU aktivitások időarányos megoszlását, valamint a különböző kernel-ek és memória műveletek végrehajtásának időtartamát.

A következő adatokat láthatjuk az eredményben:

1. GPU aktivitások:
- **`92.87%`**: A **GPU aktivitások összes idejének 92.87%-át** a `vector_add` kernel végrehajtása tette ki. Ez azt jelzi, hogy a program fő részét a kernel futtatása képezte.
  - **`40.918ms`**: A `vector_add` kernel teljes futási ideje.
  - **`1`**: A kernel pontosan **1 alkalommal** futott.
  - **`40.918ms`**: Az egyes futások időtartama, ami azonos minden egyes kernel végrehajtás esetén (mivel csak egyszer futott).

2. **[CUDA memcpy DtoH] (Device-to-Host Memory Copy):
- **`3.68%`**: A memória másolás a GPU-ról a CPU-ra (Device-to-Host) **3.68%-ban** tette ki a teljes időt.
  - **`1.6196ms`**: Ez a másolás teljes időtartama.
  - **`1`**: A memória másolás pontosan **1 alkalommal** történt.
  - **`1.6196ms`**: Az egyes memória másolások időtartama (mivel csak egyszer történt másolás).

3. **[CUDA memcpy HtoD] (Host-to-Device Memory Copy):
- **`3.45%`**: A memória másolás a CPU-ról a GPU-ra (Host-to-Device) **3.45%-ban** tette ki a teljes időt.
  - **`1.5220ms`**: A memória másolás teljes időtartama (összesen két másolás történt).
  - **`2`**: A memória másolás **2 alkalommal** történt.
  - **`761.00us`**: Az egyes másolások átlagos időtartama (768us és 743us között ingadozott).
  - **`743.66us`** és **`778.35us`**: A másolások közötti időeltérés, tehát az időintervallumok minimális és maximális értékei.

Összegzés:
- **A `vector_add` kernel** a legnagyobb részét tette ki a program futásának, és **40.918 ms**-ot igényelt.
- **Memória másolás**:
  - A **GPU-ról a CPU-ra történő másolás** (`[CUDA memcpy DtoH]`) körülbelül **1.6196 ms**-ot vett igénybe.
  - A **CPU-ról a GPU-ra történő másolás** (`[CUDA memcpy HtoD]`) összesen **1.5220 ms** időt vett igénybe, amelyet két különböző másolás hajtott végre.

Itt is látjuk, a legnagyobb időt a **kernel végrehajtása** tette ki, míg a memória másolás (GPU és CPU között) viszonylag kis időt vett igénybe az összes futási időhöz képest.


---

## Általános optimalizálási tippek
- Memóriahasználat csökkentése: Ha a memória másolás időigényes, próbáljunk meg kevesebb adatot mozgatni.  
- Regiszterhasználat figyelése: Ha túl sok a regiszter, kevesebb thread tud futni párhuzamosan.  
- Optimalizált grid/block méretek: Ha a kihasználtság alacsony, próbáljunk meg más **blockSize** értékeket.  
- Ha az L2 cache hit rate alacsony, érdemes **megosztott memóriát (shared memory)** használni a teljesítmény növelése érdekében.


## nvprof összegzés
| **Parancs** | **Mit csinál?** |
|-------------|----------------|
| `nvprof ./program` | Alapvető profilozás, futási idők elemzése |
| `nvprof --print-gpu-trace ./program` | Részletes GPU műveletek megjelenítése |
| `nvprof --print-gpu-summary ./program` | Memóriahasználat és kihasználtság elemzése |

Az nvprof egy egyszerű, de erőteljes parancssori eszköz a CUDA-programok teljesítményének mérésére. Ha mélyebb elemzést szeretnénk, az **Nsight Systems és Nsight Compute** további részleteket adhat.


# Feladatok

1. Írjunk egy CUDA programot, amely véletlenszerű pontokat generál az egységnégyzetben \([0,1] \times [0,1]\), majd kiszámolja a pontok távolságát az origótól.  
  Lépések:
  - Minden CUDA szál generáljon egy véletlen \( (x, y) \) pontot a \([0,1]\) tartományban.  
  - Számoljuk ki a pontok távolságát az origótól:  
   $d = \sqrt{x^2 + y^2}$
  - Tároljuk az eredményeket egy tömbben.  
  - Másoljuk vissza az adatokat a CPU-ra, és számoljuk ki az átlagos távolságot.  

  Az átlagos távolságnak közel kell lennie az elméleti értékhez:  $E[d] \approx 0.521405$

2. Írjunk egy szekvenciális majd egy CUDA programot mely Monte Carlo módszer segítségével, ![](https://render.githubusercontent.com/render/math?math=$n$) véletlenszerű pontot generálva, megközelítően kiszámítja a ![](https://render.githubusercontent.com/render/math?math=$\pi$) értékét. Az ![](https://render.githubusercontent.com/render/math?math=$n$) a program paramétere.

  A módszer leírása:
    
  - [http://nagysandor.eu/physlet/applets/iter1.html](http://nagysandor.eu/physlet/applets/iter1.html)

  - [http://www.tldp.org/HOWTO/Parallel-Processing-HOWTO-2.html](http://www.tldp.org/HOWTO/Parallel-Processing-HOWTO-2.html)

3. A tanult CUDA időmérést felhasználva számoljuk ki CUDA-programjaink gyorsulását. Ehhez természetesen szükség van a szekvenciális verziókra is, amelyek a hoston futnak.  

  A hoston az időmérést a `chrono::high_resolution_clock` segítségével végezzük, míg a CUDA-kernélek esetében a `cudaEventCreate` és `cudaEventRecord` függvényeket használjuk.

4. Profilozzuk a megírt programokat. Vizsgáljuk meg a kernelek futási idejét, a regiszterek és memóriahasználatot, majd az észrevételeket egy rövid beszámolóban rögzítsük (ez lehet külön fájl vagy egy új "text" cella a notebook-ban).


In [4]:
%%writefile elsofeladat.cu
#include <stdio.h>
#include <stdlib.h>
#include <cuda.h>
#include <curand_kernel.h>
#include <math.h>
#include <chrono>

using namespace std;

__global__ void generate_points_and_distances(float *distances, int n, unsigned long long seed) {
    int idx = threadIdx.x + blockIdx.x * blockDim.x;

    if (idx < n) {
        curandState state;
        curand_init(seed, idx, 0, &state);

        float x = curand_uniform(&state);
        float y = curand_uniform(&state);

        float distance = sqrtf(x * x + y * y);

        distances[idx] = distance;
    }
}

int main() {
    const int n = 1000000;  // 1 millió pont
    const int blockSize = 256;
    const int numBlocks = (n + blockSize - 1) / blockSize;
    const size_t size = n * sizeof(float);

    printf("CUDA Véletlenszerű Pontok Távolságszámítás\n");
    printf("=========================================\n");
    printf("Pontok száma: %d\n", n);
    printf("Block méret: %d\n", blockSize);
    printf("Blokkok száma: %d\n", numBlocks);
    printf("Elméleti átlagos távolság: ~0.521405\n\n");

    float *h_distances = (float*)malloc(size);
    if (h_distances == NULL) {
        printf("Hiba: Host memória foglalás sikertelen!\n");
        return -1;
    }

    float *d_distances;
    cudaError_t cudaStatus = cudaMalloc(&d_distances, size);
    if (cudaStatus != cudaSuccess) {
        printf("Hiba: Device memória foglalás sikertelen!\n");
        free(h_distances);
        return -1;
    }

    cudaEvent_t start, stop;
    cudaEventCreate(&start);
    cudaEventCreate(&stop);

    auto cpu_start = chrono::high_resolution_clock::now();

    cudaEventRecord(start, 0);

    generate_points_and_distances<<<numBlocks, blockSize>>>(d_distances, n, 12345ULL);

    cudaStatus = cudaDeviceSynchronize();
    if (cudaStatus != cudaSuccess) {
        printf("Hiba: Kernel végrehajtás sikertelen! %s\n", cudaGetErrorString(cudaStatus));
        cudaFree(d_distances);
        free(h_distances);
        return -1;
    }

    cudaEventRecord(stop, 0);
    cudaEventSynchronize(stop);

    cudaStatus = cudaMemcpy(h_distances, d_distances, size, cudaMemcpyDeviceToHost);
    if (cudaStatus != cudaSuccess) {
        printf("Hiba: Memória másolás sikertelen!\n");
        cudaFree(d_distances);
        free(h_distances);
        return -1;
    }

    auto cpu_end = chrono::high_resolution_clock::now();

    double sum = 0.0;
    for (int i = 0; i < n; i++) {
        sum += h_distances[i];
    }
    double average_distance = sum / n;

    float cuda_time;
    cudaEventElapsedTime(&cuda_time, start, stop);

    auto cpu_duration = chrono::duration_cast<chrono::milliseconds>(cpu_end - cpu_start);

    printf("Eredmények:\n");
    printf("-----------\n");
    printf("Számított átlagos távolság: %.6f\n", average_distance);
    printf("Eltérés az elméleti értéktől: %.6f\n", fabs(average_distance - 0.521405));
    printf("\nTeljesítmény:\n");
    printf("CUDA kernel futási idő: %.2f ms\n", cuda_time);
    printf("Teljes CPU idő: %ld ms\n", cpu_duration.count());
    printf("Feldolgozott pontok/másodperc: %.0f\n", n / (cuda_time / 1000.0));

    float min_dist = h_distances[0];
    float max_dist = h_distances[0];
    for (int i = 1; i < n; i++) {
        if (h_distances[i] < min_dist) min_dist = h_distances[i];
        if (h_distances[i] > max_dist) max_dist = h_distances[i];
    }

    printf("\nStatisztikák:\n");
    printf("Minimális távolság: %.6f\n", min_dist);
    printf("Maximális távolság: %.6f\n", max_dist);
    printf("Tartomány: [%.6f, %.6f]\n", min_dist, max_dist);

    printf("\nMintaértékek (első 10 távolság):\n");
    for (int i = 0; i < 10; i++) {
        printf("%.6f ", h_distances[i]);
    }
    printf("\n");


    cudaStatus = cudaGetLastError();
    if (cudaStatus != cudaSuccess) {
        printf("CUDA hiba: %s\n", cudaGetErrorString(cudaStatus));
    }

    cudaEventDestroy(start);
    cudaEventDestroy(stop);
    cudaFree(d_distances);
    free(h_distances);

    return 0;
}

Overwriting elsofeladat.cu


In [5]:
!nvcc -arch=sm_75 -o elsofeladat elsofeladat.cu

In [6]:
!./elsofeladat

CUDA Véletlenszerű Pontok Távolságszámítás
Pontok száma: 1000000
Block méret: 256
Blokkok száma: 3907
Elméleti átlagos távolság: ~0.521405

Eredmények:
-----------
Számított átlagos távolság: 0.765454
Eltérés az elméleti értéktől: 0.244049

Teljesítmény:
CUDA kernel futási idő: 41.36 ms
Teljes CPU idő: 44 ms
Feldolgozott pontok/másodperc: 24175593

Statisztikák:
Minimális távolság: 0.002267
Maximális távolság: 1.413326
Tartomány: [0.002267, 1.413326]

Mintaértékek (első 10 távolság):
0.484252 0.864508 0.719313 1.323977 0.083039 0.411939 1.020200 0.805797 0.667082 0.453043 


In [18]:
%%writefile masodikfeladat.cu
#include <stdio.h>
#include <stdlib.h>
#include <cuda.h>
#include <curand_kernel.h>
#include <math.h>
#include <chrono>
#include <random>

using namespace std;

__global__ void monte_carlo_pi_kernel(unsigned long long *hits_inside, unsigned long long n_per_thread, unsigned long long seed) {
    int idx = threadIdx.x + blockIdx.x * blockDim.x;

    curandState state;
    curand_init(seed, idx, 0, &state);

    unsigned long long local_hits = 0;

    for (unsigned long long i = 0; i < n_per_thread; i++) {
        float x = curand_uniform(&state) * 2.0f - 1.0f;  // [-1, 1]
        float y = curand_uniform(&state) * 2.0f - 1.0f;  // [-1, 1]

        if (x * x + y * y <= 1.0f) {
            local_hits++;
        }
    }

    hits_inside[idx] = local_hits;
}

double sequential_monte_carlo_pi(unsigned long long n) {

    auto start_time = chrono::high_resolution_clock::now();

    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<double> dis(-1.0, 1.0);

    unsigned long long hits_inside = 0;

    for (unsigned long long i = 0; i < n; i++) {
        double x = dis(gen);
        double y = dis(gen);

        if (x * x + y * y <= 1.0) {
            hits_inside++;
        }
    }

    auto end_time = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end_time - start_time);

    double pi_estimate = 4.0 * hits_inside / n;

    printf("Szekvenciális eredmények:\n");
    printf("Pontok száma: %llu\n", n);
    printf("Körön belüli pontok: %llu\n", hits_inside);
    printf("π becslés: %.10f\n", pi_estimate);
    printf("Eltérés a valós π-től: %.10f\n", fabs(pi_estimate - M_PI));
    printf("Futási idő: %ld ms\n", duration.count());
    printf("Pontok/másodperc: %.0f\n\n", n / (duration.count() / 1000.0));

    return pi_estimate;
}

double cuda_monte_carlo_pi(unsigned long long n, int blockSize) {

    int numBlocks = min(65535, (int)((n + blockSize - 1) / blockSize));
    unsigned long long points_per_thread = (n + numBlocks * blockSize - 1) / (numBlocks * blockSize);
    unsigned long long total_points = numBlocks * blockSize * points_per_thread;

    printf("Block méret: %d\n", blockSize);
    printf("Blokkok száma: %d\n", numBlocks);
    printf("Pontok szálanként: %llu\n", points_per_thread);
    printf("Összes feldolgozott pont: %llu\n", total_points);

    unsigned long long *d_hits;
    size_t hits_size = numBlocks * blockSize * sizeof(unsigned long long);
    cudaError_t cudaStatus = cudaMalloc(&d_hits, hits_size);
    if (cudaStatus != cudaSuccess) {
        printf("Hiba: Device memória foglalás sikertelen!\n");
        return -1;
    }

    cudaEvent_t start, stop;
    cudaEventCreate(&start);
    cudaEventCreate(&stop);

    cudaEventRecord(start, 0);

    monte_carlo_pi_kernel<<<numBlocks, blockSize>>>(d_hits, points_per_thread, 42ULL);

    cudaEventRecord(stop, 0);
    cudaEventSynchronize(stop);

    cudaStatus = cudaDeviceSynchronize();
    if (cudaStatus != cudaSuccess) {
        printf("Hiba: Kernel végrehajtás sikertelen! %s\n", cudaGetErrorString(cudaStatus));
        cudaFree(d_hits);
        return -1;
    }

    unsigned long long *h_hits = (unsigned long long*)malloc(hits_size);
    cudaStatus = cudaMemcpy(h_hits, d_hits, hits_size, cudaMemcpyDeviceToHost);
    if (cudaStatus != cudaSuccess) {
        printf("Hiba: Memória másolás sikertelen!\n");
        cudaFree(d_hits);
        free(h_hits);
        return -1;
    }

    unsigned long long total_hits = 0;
    for (int i = 0; i < numBlocks * blockSize; i++) {
        total_hits += h_hits[i];
    }

    float cuda_time;
    cudaEventElapsedTime(&cuda_time, start, stop);

    double pi_estimate = 4.0 * total_hits / total_points;

    printf("CUDA eredmények:\n");
    printf("Feldolgozott pontok: %llu\n", total_points);
    printf("Körön belüli pontok: %llu\n", total_hits);
    printf("π becslés: %.10f\n", pi_estimate);
    printf("Eltérés a valós π-től: %.10f\n", fabs(pi_estimate - M_PI));
    printf("CUDA kernel futási idő: %.2f ms\n", cuda_time);
    printf("Pontok/másodperc: %.0f\n\n", total_points / (cuda_time / 1000.0));

    cudaEventDestroy(start);
    cudaEventDestroy(stop);
    cudaFree(d_hits);
    free(h_hits);

    return pi_estimate;
}

int main(int argc, char** argv) {
    printf("Monte Carlo π Számítás - Szekvenciális vs CUDA\n");
    printf("===============================================\n");
    printf("Valós π érték: %.10f\n\n", M_PI);

    unsigned long long n = 10000000;
    int blockSize = 256;

    if (argc >= 2) {
        n = strtoull(argv[1], NULL, 10);
        if (n == 0) {
            printf("Hiba: Érvénytelen pontszám!\n");
            printf("Használat: %s <pontok_száma> [block_méret]\n", argv[0]);
            return -1;
        }
    }

    if (argc >= 3) {
        blockSize = atoi(argv[2]);
        if (blockSize <= 0 || blockSize > 1024) {
            printf("Hiba: Érvénytelen block méret! (1-1024 között)\n");
            return -1;
        }
    }

    printf("Paraméterek:\n");
    printf("Pontok száma: %llu\n", n);
    printf("CUDA block méret: %d\n\n", blockSize);

    auto seq_start = chrono::high_resolution_clock::now();
    double pi_sequential = sequential_monte_carlo_pi(n);
    auto seq_end = chrono::high_resolution_clock::now();
    auto seq_duration = chrono::duration_cast<chrono::milliseconds>(seq_end - seq_start);

    auto cuda_start = chrono::high_resolution_clock::now();
    double pi_cuda = cuda_monte_carlo_pi(n, blockSize);
    auto cuda_end = chrono::high_resolution_clock::now();
    auto cuda_duration = chrono::duration_cast<chrono::milliseconds>(cuda_end - cuda_start);

    printf("Összehasonlítás:\n");
    printf("================\n");
    printf("Szekvenciális π: %.10f (eltérés: %.10f)\n", pi_sequential, fabs(pi_sequential - M_PI));
    printf("CUDA π:          %.10f (eltérés: %.10f)\n", pi_cuda, fabs(pi_cuda - M_PI));
    printf("Különbség:       %.10f\n\n", fabs(pi_sequential - pi_cuda));

    printf("Teljesítmény:\n");
    printf("=============\n");
    printf("Szekvenciális teljes idő: %ld ms\n", seq_duration.count());
    printf("CUDA teljes idő:          %ld ms\n", cuda_duration.count());

    if (seq_duration.count() > 0 && cuda_duration.count() > 0) {
        double speedup = (double)seq_duration.count() / cuda_duration.count();
        printf("Gyorsítás:                %.2fx\n", speedup);

        printf("Elméleti hatékonyság:     %.2f%%\n", (speedup / 1000.0) * 100); // Példa 1000 core GPU-ra
    }

    printf("\nPrecizitás elemzés:\n");
    printf("==================\n");
    double theoretical_error = 1.96 / sqrt(n);  // 95% konfidencia intervallum
    printf("Elméleti hiba (95%% CI):   ±%.6f\n", theoretical_error);
    printf("Szekvenciális hiba:       %.6f\n", fabs(pi_sequential - M_PI));
    printf("CUDA hiba:                %.6f\n", fabs(pi_cuda - M_PI));

    if (fabs(pi_sequential - M_PI) > theoretical_error * 2) {
        printf("- Növelje a mintavételezés számát a jobb pontosságért\n");
    }
    if (seq_duration.count() < cuda_duration.count()) {
        printf("- Kis mintaszámnál a CUDA overhead nagyobb lehet\n");
        printf("- Próbáljon nagyobb pontszámot (pl. %llu)\n", n * 10);
    }

    return 0;
}

Overwriting masodikfeladat.cu


In [21]:
!nvcc -arch=sm_75 -o masodikfeladat masodikfeladat.cu

In [22]:
!./masodikfeladat

Monte Carlo π Számítás - Szekvenciális vs CUDA
Valós π érték: 3.1415926536

Paraméterek:
Pontok száma: 10000000
CUDA block méret: 256

Szekvenciális eredmények:
Pontok száma: 10000000
Körön belüli pontok: 7853511
π becslés: 3.1414044000
Eltérés a valós π-től: 0.0001882536
Futási idő: 1373 ms
Pontok/másodperc: 7283321

Block méret: 256
Blokkok száma: 39063
Pontok szálanként: 1
Összes feldolgozott pont: 10000128
CUDA eredmények:
Feldolgozott pontok: 10000128
Körön belüli pontok: 7851674
π becslés: 3.1406293999
Eltérés a valós π-től: 0.0009632536
CUDA kernel futási idő: 321.59 ms
Pontok/másodperc: 31096167

Összehasonlítás:
Szekvenciális π: 3.1414044000 (eltérés: 0.0001882536)
CUDA π:          3.1406293999 (eltérés: 0.0009632536)
Különbség:       0.0007750001

Teljesítmény:
Szekvenciális teljes idő: 1373 ms
CUDA teljes idő:          601 ms
Gyorsítás:                2.28x
Elméleti hatékonyság:     0.23%

Precizitás elemzés:
Elméleti hiba (95% CI):   ±0.000620
Szekvenciális hiba:       0.00

In [23]:
!nvprof ./elsofeladat

CUDA Véletlenszerű Pontok Távolságszámítás
Pontok száma: 1000000
Block méret: 256
Blokkok száma: 3907
Elméleti átlagos távolság: ~0.521405

==10168== NVPROF is profiling process 10168, command: ./elsofeladat
Eredmények:
-----------
Számított átlagos távolság: 0.765454
Eltérés az elméleti értéktől: 0.244049

Teljesítmény:
CUDA kernel futási idő: 41.43 ms
Teljes CPU idő: 44 ms
Feldolgozott pontok/másodperc: 24138525

Statisztikák:
Minimális távolság: 0.002267
Maximális távolság: 1.413326
Tartomány: [0.002267, 1.413326]

Mintaértékek (első 10 távolság):
0.484252 0.864508 0.719313 1.323977 0.083039 0.411939 1.020200 0.805797 0.667082 0.453043 
==10168== Profiling application: ./elsofeladat
==10168== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   96.24%  40.872ms         1  40.872ms  40.872ms  40.872ms  generate_points_and_distances(float*, int, __int64)
                    3.76%  1.5968ms         1  1.5968ms  1.5968ms 

In [24]:
!nvprof --print-gpu-trace ./elsofeladat

CUDA Véletlenszerű Pontok Távolságszámítás
Pontok száma: 1000000
Block méret: 256
Blokkok száma: 3907
Elméleti átlagos távolság: ~0.521405

==10275== NVPROF is profiling process 10275, command: ./elsofeladat
Eredmények:
-----------
Számított átlagos távolság: 0.765454
Eltérés az elméleti értéktől: 0.244049

Teljesítmény:
CUDA kernel futási idő: 41.27 ms
Teljes CPU idő: 44 ms
Feldolgozott pontok/másodperc: 24231681

Statisztikák:
Minimális távolság: 0.002267
Maximális távolság: 1.413326
Tartomány: [0.002267, 1.413326]

Mintaértékek (első 10 távolság):
0.484252 0.864508 0.719313 1.323977 0.083039 0.411939 1.020200 0.805797 0.667082 0.453043 
==10275== Profiling application: ./elsofeladat
==10275== Profiling result:
   Start  Duration            Grid Size      Block Size     Regs*    SSMem*    DSMem*      Size  Throughput  SrcMemType  DstMemType           Device   Context    Stream  Name
318.80ms  40.793ms           (3907 1 1)       (256 1 1)        63        0B        0B         -       

In [25]:
!nvprof --print-gpu-summary ./elsofeladat

CUDA Véletlenszerű Pontok Távolságszámítás
Pontok száma: 1000000
Block méret: 256
Blokkok száma: 3907
Elméleti átlagos távolság: ~0.521405

==10364== NVPROF is profiling process 10364, command: ./elsofeladat
Eredmények:
-----------
Számított átlagos távolság: 0.765454
Eltérés az elméleti értéktől: 0.244049

Teljesítmény:
CUDA kernel futási idő: 41.38 ms
Teljes CPU idő: 44 ms
Feldolgozott pontok/másodperc: 24163704

Statisztikák:
Minimális távolság: 0.002267
Maximális távolság: 1.413326
Tartomány: [0.002267, 1.413326]

Mintaértékek (első 10 távolság):
0.484252 0.864508 0.719313 1.323977 0.083039 0.411939 1.020200 0.805797 0.667082 0.453043 
==10364== Profiling application: ./elsofeladat
==10364== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   96.23%  40.863ms         1  40.863ms  40.863ms  40.863ms  generate_points_and_distances(float*, int, __int64)
                    3.77%  1.6023ms         1  1.6023ms  1.6023ms 

In [26]:
!nvprof ./masodikfeladat

Monte Carlo π Számítás - Szekvenciális vs CUDA
Valós π érték: 3.1415926536

Paraméterek:
Pontok száma: 10000000
CUDA block méret: 256

Szekvenciális eredmények:
Pontok száma: 10000000
Körön belüli pontok: 7853911
π becslés: 3.1415644000
Eltérés a valós π-től: 0.0000282536
Futási idő: 1373 ms
Pontok/másodperc: 7283321

Block méret: 256
Blokkok száma: 39063
Pontok szálanként: 1
Összes feldolgozott pont: 10000128
==10457== NVPROF is profiling process 10457, command: ./masodikfeladat
CUDA eredmények:
Feldolgozott pontok: 10000128
Körön belüli pontok: 7851674
π becslés: 3.1406293999
Eltérés a valós π-től: 0.0009632536
CUDA kernel futási idő: 306.94 ms
Pontok/másodperc: 32579695

Összehasonlítás:
Szekvenciális π: 3.1415644000 (eltérés: 0.0000282536)
CUDA π:          3.1406293999 (eltérés: 0.0009632536)
Különbség:       0.0009350001

Teljesítmény:
Szekvenciális teljes idő: 1373 ms
CUDA teljes idő:          713 ms
Gyorsítás:                1.93x
Elméleti hatékonyság:     0.19%

Precizitás elem

In [27]:
!nvprof --print-gpu-trace ./masodikfeladat

Monte Carlo π Számítás - Szekvenciális vs CUDA
Valós π érték: 3.1415926536

Paraméterek:
Pontok száma: 10000000
CUDA block méret: 256

Szekvenciális eredmények:
Pontok száma: 10000000
Körön belüli pontok: 7852649
π becslés: 3.1410596000
Eltérés a valós π-től: 0.0005330536
Futási idő: 1374 ms
Pontok/másodperc: 7278020

Block méret: 256
Blokkok száma: 39063
Pontok szálanként: 1
Összes feldolgozott pont: 10000128
==10528== NVPROF is profiling process 10528, command: ./masodikfeladat
CUDA eredmények:
Feldolgozott pontok: 10000128
Körön belüli pontok: 7851674
π becslés: 3.1406293999
Eltérés a valós π-től: 0.0009632536
CUDA kernel futási idő: 324.28 ms
Pontok/másodperc: 30837963

Összehasonlítás:
Szekvenciális π: 3.1410596000 (eltérés: 0.0005330536)
CUDA π:          3.1406293999 (eltérés: 0.0009632536)
Különbség:       0.0004302001

Teljesítmény:
Szekvenciális teljes idő: 1374 ms
CUDA teljes idő:          717 ms
Gyorsítás:                1.92x
Elméleti hatékonyság:     0.19%

Precizitás elem

In [28]:
!nvprof --print-gpu-summary ./masodikfeladat

Monte Carlo π Számítás - Szekvenciális vs CUDA
Valós π érték: 3.1415926536

Paraméterek:
Pontok száma: 10000000
CUDA block méret: 256

Szekvenciális eredmények:
Pontok száma: 10000000
Körön belüli pontok: 7853518
π becslés: 3.1414072000
Eltérés a valós π-től: 0.0001854536
Futási idő: 1897 ms
Pontok/másodperc: 5271481

Block méret: 256
Blokkok száma: 39063
Pontok szálanként: 1
Összes feldolgozott pont: 10000128
==10615== NVPROF is profiling process 10615, command: ./masodikfeladat
CUDA eredmények:
Feldolgozott pontok: 10000128
Körön belüli pontok: 7851674
π becslés: 3.1406293999
Eltérés a valós π-től: 0.0009632536
CUDA kernel futási idő: 348.69 ms
Pontok/másodperc: 28679490

Összehasonlítás:
Szekvenciális π: 3.1414072000 (eltérés: 0.0001854536)
CUDA π:          3.1406293999 (eltérés: 0.0009632536)
Különbség:       0.0007778001

Teljesítmény:
Szekvenciális teljes idő: 1897 ms
CUDA teljes idő:          785 ms
Gyorsítás:                2.42x
Elméleti hatékonyság:     0.24%

Precizitás elem